### PR TITLE
Bidirectional language model + fixes to the end-to-end tests

### PR DESCRIPTION
**Pull request by sebastianruder**
This is a PR for adding a bidirectional LM as discussed in https://github.com/n-waves/ulmfit-multilingual/issues/4.

### CODE REVIEW

File: fastai_contrib/models.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        new_hidden,raw_outputs,outputs = [],[],[]
+        for l, (rnn,hid_dp) in enumerate(zip(self.forward_rnns, self.hidden_dps)):
+            raw_output, new_h = rnn(raw_output, self.hidden[l])
+            new_hidden.append(new_h)
+            raw_outputs.append(raw_output)
+            if l != self.n_layers - 1: raw_output = hid_dp(raw_output)
+            outputs.append(raw_output)
+        self.hidden = to_detach(new_hidden)
+        return raw_outputs, outputs
+
```
[the remark this answers is not in the document]
Comment (dust0x) [reply fastai_contrib/models.py]: This should, in fact, start after line 60.

File: fastai_contrib/models.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        raw_output = self.input_dp(self.encoder_dp(input))
+
+        # TODO get reverse input and compute backward representation
+        new_hidden,raw_outputs,outputs = [],[],[]
+        for l, (rnn,hid_dp) in enumerate(zip(self.forward_rnns, self.hidden_dps)):
+            raw_output, new_h = rnn(raw_output, self.hidden[l])
+            new_hidden.append(new_h)
+            raw_outputs.append(raw_output)
+            if l != self.n_layers - 1: raw_output = hid_dp(raw_output)
+            outputs.append(raw_output)
```
Comment (dust0x) [inline fastai_contrib/models.py]:
```suggestion
            outputs.append(raw_output)
                
        for l, (rnn,hid_dp) in enumerate(zip(self.backward_rnns, self.hidden_dps)):
            raw_output, new_h = rnn(raw_output_backward, self.hidden['backward'][l])
            new_hidden.append(new_h)
            raw_outputs.append(raw_output)
            if l != self.n_layers - 1: raw_output = hid_dp(raw_output)
            outputs.append(raw_output)
```

File: fastai_contrib/models.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        self.encoder.weight.data.uniform_(-self.initrange, self.initrange)
+        self.input_dp = RNNDropout(input_p)
+        self.hidden_dps = nn.ModuleList([RNNDropout(hidden_p) for l in range(n_layers)])
+
+    def forward(self, input:LongTensor)->Tuple[Tensor,Tensor]:
+        sl,bs = input.size()
+        if bs!=self.bs:
+            self.bs=bs
+            self.reset()
+        raw_output = self.input_dp(self.encoder_dp(input))
```
[the remark this answers is not in the document]
Comment (sebastianruder) [reply fastai_contrib/models.py]: Thanks for the suggestion! These suggested changes are super useful! Negative indexing does not work in Pytorch yet unfortunately. :-/ So we'd either have to convert the tensor to numpy or do the reversion before.

[the remark this answers is not in the document]
Comment (PiotrCzapla) [reply fastai_contrib/models.py]:
```
t=torch.LongTensor([10,20,30,40,50]);print(t)
idx=torch.arange(t.shape[0]-1,-1,-1)
print(t.index_select(0,idx))
```

And unfortunately [::-1] would not work anyway as input is multidimensional at this stage.

[the remark this answers is not in the document]
Comment (dust0x) [reply fastai_contrib/models.py]: Yes, I'm sorry I realised this only after commenting. I thought of letting it stay as a sketch of what could go in its place.

[the remark this answers is not in the document]
Comment (dust0x) [reply fastai_contrib/models.py]: In any case this (and some of the other stuff) isn't required if we're not doing the reversing inside `forward` function anyway.

**Comment by PiotrCzapla** [comment]
Guys, I will merge the changes in bilm branch as the current version of master branch does not work there are few bugs that I've corrected in this branch.

**Pull request merged**